src/parser/deserializer.ts

```diff
@@ -169,12 +176,17 @@ function deserializeObject(
       )
         throw new Error('bad string length in bson');
 
-      if (!validateUtf8(buffer, index, index + stringSize - 1)) {
-        throw new Error('Invalid UTF-8 string in BSON document');
-      }
-
       value = buffer.toString('utf8', index, index + stringSize - 1);
 
+      for (let i = 0; i < value.length; i++) {
+        if (value.charCodeAt(i) === 0xfffd) {
+          if (!validateUtf8(buffer, index, index + stringSize - 1)) {
+            throw new Error('Invalid UTF-8 string in BSON document');
+          }
```

Comment:
We might change some stuff here depending on the other work so lets leave addressing this for last

---

Is there a speed up here if we just look for the `0xFF, 0xFD` sequence in the original bytes and if it isn't present but it is present in the output then its invalid utf8? vs actually validating the utf8